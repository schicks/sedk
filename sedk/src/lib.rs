#![allow(dead_code)]
mod dsl;
pub use dsl::{
    primitives::IntoFields,
    field::{Field, FieldType, Indexable, IndexMapping},
    analysis::{Analyzer, Normalizer}
};