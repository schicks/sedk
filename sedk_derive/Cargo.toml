[package]
name = "sedk_derive"
version = "0.1.0"
authors = ["Sam Schick <schicksw@protonmail.com>"]
edition = "2018"

[lib]
proc-macro = true

[[test]]
name = "tests"
path = "tests/runner.rs"

[dependencies]
proc-macro2 = "1.0.26"
quote = "1.0.9"
syn = "1.0.69"
trybuild = "1.0.42"

[dev-dependencies]
sedk = { path = "../sedk"}
