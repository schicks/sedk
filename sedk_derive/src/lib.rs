/// # Examples
/// ```rust
/// use sedk::{IntoFields, Indexable, IndexMapping, Field, FieldType, Analyzer, Normalizer};
/// use sedk_derive::{IntoFields, Indexable};

/// #[derive(IntoFields)]
/// pub struct Child {
///     integer: i32,
///     string: String
/// }
///
/// #[derive(IntoFields, Indexable)]
/// pub struct Parent {
///     nested: Child
/// }

/// fn main() {
///     let fields = Child::into_fields(None);
///     assert_eq!(
///         fields,
///         vec![
///             Field {
///                 name: "integer".to_owned(), 
///                 field_type: FieldType::Integer, 
///                 fields: Vec::new()
///             },
///             Field {
///                 name: "string".to_owned(),
///                 field_type: FieldType::Text {analyzer: Analyzer::default()},
///                 fields: vec![
///                     Field {
///                         name: "string.keyword".to_owned(),
///                         field_type: FieldType::Keyword {normalizer: Normalizer::default()},
///                         fields: Vec::new()
///                     }
///                 ]
///             }
///         ]
///     );
///     let parent_mapping = Parent::index_mapping();
///     assert_eq!(
///         parent_mapping.fields,
///         vec![
///             Field {
///                 name: "nested.integer".to_owned(), 
///                 field_type: FieldType::Integer, 
///                 fields: Vec::new()
///             },
///             Field {
///                 name: "nested.string".to_owned(),
///                 field_type: FieldType::Text {analyzer: Analyzer::default()},
///                 fields: vec![
///                     Field {
///                         name: "nested.string.keyword".to_owned(),
///                         field_type: FieldType::Keyword {normalizer: Normalizer::default()},
///                         fields: Vec::new()
///                     }
///                 ]
///             }
///         ]
///     );
/// }
/// ```
mod indexable;
mod into_fields;
use indexable::impl_indexable;
use into_fields::impl_into_fields;

use proc_macro::TokenStream;
use syn::{parse_macro_input, DeriveInput};

#[proc_macro_derive(Indexable)]
pub fn indexable(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as DeriveInput);
    impl_indexable(input).into()
}

#[proc_macro_derive(IntoFields)]
pub fn into_fields(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as DeriveInput);
    impl_into_fields(input).into()
}