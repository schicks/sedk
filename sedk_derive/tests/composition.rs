use sedk::{IntoFields, Field, FieldType, Analyzer, Normalizer};
use sedk_derive::IntoFields;

#[derive(IntoFields)]
pub struct Child {
    integer: i32,
    string: String
}

#[derive(IntoFields)]
pub struct Parent {
    child: Child
}

fn main() {
    assert_eq!(
        Parent::into_fields(None),
        vec![
            Field {
                name: "child.integer".to_owned(), 
                field_type: FieldType::Integer, 
                fields: Vec::new()
            },
            Field {
                name: "child.string".to_owned(),
                field_type: FieldType::Text {analyzer: Analyzer::default()},
                fields: vec![
                    Field {
                        name: "child.string.keyword".to_owned(),
                        field_type: FieldType::Keyword {normalizer: Normalizer::default()},
                        fields: Vec::new()
                    }
                ]
            }
        ]
    )
}