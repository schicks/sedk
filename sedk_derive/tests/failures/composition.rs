use sedk::{IntoFields, Field};
use sedk_derive::IntoFields;

pub struct Child {
    integer: i32,
    string: String
}

#[derive(IntoFields)]
pub struct Parent {
    child: Child
}

fn main() {
}