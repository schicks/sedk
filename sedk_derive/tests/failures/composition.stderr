error[E0599]: no function or associated item named `into_fields` found for struct `Child` in the current scope
 --> $DIR/composition.rs:9:10
  |
4 | pub struct Child {
  | ---------------- function or associated item `into_fields` not found for this
...
9 | #[derive(IntoFields)]
  |          ^^^^^^^^^^ function or associated item not found in `Child`
  |
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `into_fields`, perhaps you need to implement it:
          candidate #1: `IntoFields`
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
